File: creational/factory-method/python/factory_method.py
"""Factory Method pattern demonstrated with format-specific record parsers."""

from __future__ import annotations

import csv
import io
import json
from abc import ABC, abstractmethod
from typing import Any


class RecordParser(ABC):
    """Product interface used by the ingestion workflow."""

    @abstractmethod
    def parse(self, payload: str) -> list[dict[str, Any]]:
        """Convert a raw payload into a list of records."""


class CsvRecordParser(RecordParser):
    """Concrete product for comma-separated records."""

    def parse(self, payload: str) -> list[dict[str, Any]]:
        return [dict(row) for row in csv.DictReader(io.StringIO(payload))]


class JsonRecordParser(RecordParser):
    """Concrete product for a JSON array of objects."""

    def parse(self, payload: str) -> list[dict[str, Any]]:
        try:
            parsed: Any = json.loads(payload)
        except json.JSONDecodeError as exc:
            raise ValueError("The payload is not valid JSON") from exc

        if not isinstance(parsed, list) or not all(
            isinstance(record, dict) for record in parsed
        ):
            raise ValueError("JSON input must be an array of objects")

        return [dict(record) for record in parsed]


class IngestionJob(ABC):
    """Creator containing the stable ingestion workflow.

    create_parser is the Factory Method. The workflow does not know
    which concrete parser it receives.
    """

    def run(self, payload: str) -> list[dict[str, Any]]:
        parser = self.create_parser()
        records = parser.parse(payload)
        self._validate_records(records)
        return records

    @abstractmethod
    def create_parser(self) -> RecordParser:
        """Create the parser required by this ingestion job."""

    @staticmethod
    def _validate_records(records: list[dict[str, Any]]) -> None:
        if any(not record for record in records):
            raise ValueError("Records must not be empty")


class CsvIngestionJob(IngestionJob):
    """Concrete creator that creates a CSV parser."""

    def create_parser(self) -> RecordParser:
        return CsvRecordParser()


class JsonIngestionJob(IngestionJob):
    """Concrete creator that creates a JSON parser."""

    def create_parser(self) -> RecordParser:
        return JsonRecordParser()


def ingest(job: IngestionJob, payload: str) -> list[dict[str, Any]]:
    """Client function independent of concrete parser classes."""

    return job.run(payload)


if __name__ == "__main__":
    csv_payload = "asset_id,power_kw\nINV-01,42.5\nINV-02,38.1\n"
    json_payload = '[{"asset_id": "INV-01", "power_kw": 42.5}]'

    for job, payload in (
        (CsvIngestionJob(), csv_payload),
        (JsonIngestionJob(), json_payload),
    ):
        print(f"{job.__class__.__name__}: {ingest(job, payload)}")

File: creational/factory-method/java/FactoryMethod.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Pattern;

/**
 * Factory Method pattern demonstrated with format-specific record parsers.
 */
public final class FactoryMethod {
    private FactoryMethod() {
    }

    /**
     * Product interface used by the ingestion workflow.
     */
    public interface RecordParser {
        List<Map<String, String>> parse(String payload);
    }

    /**
     * Concrete product for comma-separated records.
     */
    public static final class CsvRecordParser implements RecordParser {
        @Override
        public List<Map<String, String>> parse(String payload) {
            return parseDelimited(payload, ',');
        }
    }

    /**
     * Concrete product for tab-separated records.
     */
    public static final class TsvRecordParser implements RecordParser {
        @Override
        public List<Map<String, String>> parse(String payload) {
            return parseDelimited(payload, '\t');
        }
    }

    /**
     * Creator containing the stable ingestion workflow.
     *
     * createParser is the Factory Method. The workflow does not know
     * which concrete parser it receives.
     */
    public abstract static class IngestionJob {
        public final List<Map<String, String>> run(String payload) {
            RecordParser parser = createParser();
            List<Map<String, String>> records = parser.parse(payload);
            validateRecords(records);
            return records;
        }

        public abstract RecordParser createParser();

        private static void validateRecords(List<Map<String, String>> records) {
            if (records.stream().anyMatch(Map::isEmpty)) {
                throw new IllegalArgumentException("Records must not be empty");
            }
        }
    }

    /**
     * Concrete creator that creates a CSV parser.
     */
    public static final class CsvIngestionJob extends IngestionJob {
        @Override
        public RecordParser createParser() {
            return new CsvRecordParser();
        }
    }

    /**
     * Concrete creator that creates a TSV parser.
     */
    public static final class TsvIngestionJob extends IngestionJob {
        @Override
        public RecordParser createParser() {
            return new TsvRecordParser();
        }
    }

    /**
     * Client function independent of concrete parser classes.
     */
    public static List<Map<String, String>> ingest(
            IngestionJob job,
            String payload
    ) {
        return job.run(payload);
    }

    private static List<Map<String, String>> parseDelimited(
            String payload,
            char delimiter
    ) {
        Objects.requireNonNull(payload, "payload must not be null");

        String trimmedPayload = payload.strip();
        if (trimmedPayload.isEmpty()) {
            return List.of();
        }

        String[] lines = trimmedPayload.split("\\R");
        String[] headers = splitLine(lines[0], delimiter);
        List<Map<String, String>> records = new ArrayList<>();

        for (int lineNumber = 1; lineNumber < lines.length; lineNumber++) {
            String[] values = splitLine(lines[lineNumber], delimiter);
            if (values.length != headers.length) {
                throw new IllegalArgumentException(
                        "Column count mismatch on line " + (lineNumber + 1)
                );
            }

            Map<String, String> record = new LinkedHashMap<>();
            for (int column = 0; column < headers.length; column++) {
                record.put(headers[column].trim(), values[column].trim());
            }
            records.add(record);
        }

        return records;
    }

    private static String[] splitLine(String line, char delimiter) {
        return line.split(Pattern.quote(String.valueOf(delimiter)), -1);
    }

    public static void main(String[] args) {
        String csvPayload = "asset_id,power_kw\nINV-01,42.5\nINV-02,38.1\n";
        String tsvPayload = "asset_id\\tpower_kw\nINV-01\\t42.5\n";

        for (IngestionJob job : List.of(
                new CsvIngestionJob(),
                new TsvIngestionJob()
        )) {
            String payload = job instanceof CsvIngestionJob
                    ? csvPayload
                    : tsvPayload;
            System.out.println(
                    job.getClass().getSimpleName() + ": "
                            + ingest(job, payload)
            );
        }
    }
}

